Add missing pytest files and hook to detect them

DavidFair opened the pull request:
**Description of work.**
- Adds a pre-commit hook from https://github.com/DavidFair/pre-commit-hooks/tree/Add_pytest_checking
- Add the various missing files from the project into their CMake lists

**To test:**
- Code Review https://github.com/mantidproject/pre-commit-hooks/pull/5
- Try running pre-commit using `pre-commit run cmake-missing-pytest-files --verbose --all-files` it should pass
- Remove (a) random file(s) from CMake lists, then re-run check they are correctly detected as missing

If you are happy with the changes:
- Approve the PR in our hooks repo
- Ping me, and I'll update the URL from my fork to the Mantid version
- Then this will be ready for merge

There is no associated issue - I thought we created an issue for missing tests, but cannot find it?

*This does not require release notes* because it's not user visible

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Add new hook for checking all pytest-files are in CMake lists

Adds a new hook (pointing to a fork currently) that all pytest files are
included in CMake lists correctly

DavidFair committed:
Add config (pointing to fork) for Pytest precommit

Adds a new precommit hook to check for missing PyTest files. These will
not be run automatically by CI unless they are added

DavidFair committed:
Add missing pytest files to associated CMakeLists

Adds missing Pytest files to their respective CMakelists.

In the case of system tests we simply remove the unit test marker, as they inherit from
mantid.systemtest and should not be using unittest directly.

DavidFair committed:
Fix various tests that weren't running

Fixes various tests by adding mocks, or removing obselete ones to get
the entire codebase working.

DavidFair committed:
Fix missing files from upsteam rebase

Some of the merges got complex, so when in doubt we favoured main and
went back in this commit to add-back the missing files

DavidFair committed:
Further fixes various tests that weren't running

DavidFair committed:
Rewrite deprecator test to not use file IO

This was racey in Windows, and adds handling where a memory mapped
buffer will work just as well.

Pasarus commented on .pre-commit-config.yaml:
```diff
@@ -13,16 +12,19 @@ repos:
 
-  - repo: https://github.com/mantidproject/pre-commit-hooks.git
-    rev: 8809bbab4c8b9609f559e2428e6255e9407c3efe
+  - repo: https://github.com/DavidFair/pre-commit-hooks.git
```
```suggestion
  - repo: https://github.com/mantidproject/pre-commit-hooks.git
```

Pasarus commented on .pre-commit-config.yaml:
```diff
@@ -13,16 +12,19 @@ repos:
-  - repo: https://github.com/mantidproject/pre-commit-hooks.git
-    rev: 8809bbab4c8b9609f559e2428e6255e9407c3efe
+  - repo: https://github.com/DavidFair/pre-commit-hooks.git
+    rev: 286a37a33b4c583808d2224735d7c461d4079773
```
```suggestion
    rev: f99fd08a5f4ece5a196adeda23d20b4491cfd97b
```

Pasarus reviewed (changes requested):
Just a quick change to use the merged hook opposed to your fork version.

DavidFair committed:
Point to merged commit hook in upstream repo

Now the work has been merged from the fork, we can point to the main
repo for production workloads instead

gemmaguest commented:
@DavidFair Just wondering why some existing tests have been changed/deleted?

gemmaguest commented on Framework/PythonInterface/test/python/plugins/functions/RedfieldTest.py:
```diff
@@ -19,7 +19,7 @@ def test_function_has_been_registered(self):
 	def test_function_output(self):
 		input = [0.0, 4.0, 8.0, 12.0]
-		expected = [0.2, -0.05237002010607009, 0.061888393844228465, -0.03887640266005458]
+		expected = [7.34780775e-08, 7.34759179e-08, 7.34694400e-08, 7.34586461e-08]
```
These values look strange given that the tolerance is e-05?

gemmaguest replied:
@DavidFair Did you find out why these values had changed?

gemmaguest commented on Framework/PythonInterface/test/python/plugins/algorithms/IndexPeaksTest.py:
```diff
@@ -19,22 +19,13 @@ class IndexPeaksTest(unittest.TestCase):
         IndexPeaks(PeaksWorkspace="test", Tolerance=0.12)
 
-        # testing zero (default) modVec with maxOrder=1
-        IndexPeaks(
```
Is it worth checking if we can get these tests working? Maybe as a future issue if that's difficult - it would still be good to get the rest of these tests in for now seeing as none were being run till now.

DavidFair replied:
@RichardWaiteSTFC This fails now as something in the past has changed, do you know if this algorithm call need tweaking / is still valid?

RichardWaiteSTFC replied:
The call to IndexPeaks looks OK to me - don't think anything has changed recently. The last thing I can find was one of mine from August (just changing a warning).
https://github.com/mantidproject/mantid/commit/11886ae3a86c923246d3222829358702e05b3a86#diff-512d369be72489a814cd4017a6c21a497033e2e8a0ad2845097bb0d05d0a4231

RichardWaiteSTFC replied:
Ah I think the error is because a modulation vector hasn't been provided - let me check to see what the expected behaviour is

gemmaguest commented on qt/python/mantidqtinterfaces/test/Muon/PlottingView_test.py:
```diff
@@ -1,420 +0,0 @@
-# Mantid Repository : https://github.com/mantidproject/mantid
```
It looks like this code has been removed but maybe we should double check with @AnthonyLim23 that it hasn't been moved elsewhere?

AnthonyLim23 replied:
It should be there https://github.com/mantidproject/mantid/blob/main/qt/python/mantidqtinterfaces/test/Muon/PlottingView_test.py

DavidFair replied:
But `ElementalAnalysis/Plotting` (which is imports) is not a file or module in the project?

AnthonyLim23 replied:
this is from the first attempt of a separate plotting widget, which is also in Python. We needed to test the integration. As this is specific to muons it was decided that this was the best place for it.

AnthonyLim23 replied:
Looking into this, it seems like it was an unreliable test and was meant to have been deleted https://github.com/mantidproject/mantid/commit/fc0189ae6586d2e3f93d9bedfe6fd8ea18669059#diff-8ffab017c1c6a613568c272b90222401d627d251b1482b39feb82c78b9d9c4c3R40

gemmaguest commented on qt/widgets/common/test/FileFinderWidgetTest.py:
```diff
@@ -1,50 +0,0 @@
-# Mantid Repository : https://github.com/mantidproject/mantid
```
This is just testing qt4 stuff so I'm happy with this being removed :+1:

gemmaguest commented on scripts/test/DirectPropertyManater_OldInterfaceXest.py:
```diff
@@ -1,330 +0,0 @@
-# Mantid Repository : https://github.com/mantidproject/mantid
```
This has been disabled, had typos in the name and not been meaningfully changed since 2015. There's also another test with the same name that is more recent so I think I'm happy with this being removed :+1:

gemmaguest commented on scripts/test/SANS/state/StateBuilderTest.py:
```diff
@@ -1,65 +0,0 @@
-# Mantid Repository : https://github.com/mantidproject/mantid
```
State builder was removed from SANS so happy for this to be removed :+1:

gemmaguest reviewed (changes requested):
Thanks @DavidFair for going through these with me. Lots of good tidying here and removing of obsolete tests. Good work :+1: There are just a couple I think we should double check. See in-line comments.

AnthonyLim23 commented on qt/python/mantidqtinterfaces/test/Muon/PlottingView_test.py:
```diff
@@ -1,420 +0,0 @@
-from unittest import mock
-from mantidqt.utils.qt.testing import start_qapplication
-
-from mantidqtinterfaces.Muon.GUI.ElementalAnalysis.Plotting.subPlot_object import subPlot
```
https://github.com/mantidproject/mantid/tree/main/qt/python/mantidqtinterfaces/mantidqtinterfaces/MultiPlotting

AnthonyLim23 replied:
should it be in here https://github.com/mantidproject/mantid/tree/main/qt/python/mantidqtinterfaces/test/MultiPlotting

RichardWaiteSTFC committed:
Refactor IndexPeaks python binding tests

DavidFair committed:
Merge pull request #1 from mantidproject/IndexPEaks_test_for_David

Refactor IndexPeaks python binding tests

DavidFair committed:
Merge remote-tracking branch 'upstream/main' into Add_missing_pytest_files

Perform a manual merge instead of a rebase to avoid rewriting history
with multiple authors

DavidFair committed:
Fix garbage RedField test input(s)

- Refactor RedField to use a built in named constant
- Use numpy.power instead of **2

For the test we were using a tiny Gauss value. The original value was
0.1, for reference a Iron Bar Magnet has a value of 100G so naturally
we're going to get garbage back out.

Tweak the values in to be more sensible as they were falling off to 0
due to the 1/x^2 calculation

DavidFair commented:
@AnthonyLim23 Could you double check https://github.com/mantidproject/mantid/pull/32592/commits/cc719fcbafb33af618d51e385978989a1a43a856

gemmaguest commented:
Great, thanks everyone :+1: @Pasarus could you re-approve if you're happy with the latest changes?
